consensus/scheme/rolldpos/chainmanager.go

```diff
@@ -0,0 +1,367 @@
+	// create a new block
+	log.L().Debug("Produce a new block.", zap.Uint64("height", newblockHeight), zap.Time("timestamp", timestamp), log.Hex("prevHash", tipHash[:]))
+	// run execution and update state trie root hash
+	blk, err := fc.bbf.Mint(ctx, pk)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to create block")
+	}
+	if err = fc.pool.AddBlock(blk); err != nil {
+		blkHash := blk.HashBlock()
+		log.L().Error("failed to add proposal", zap.Error(err), zap.Uint64("height", blk.Height()), log.Hex("hash", blkHash[:]))
+	}
+	blockMtc.WithLabelValues("MintGas").Set(float64(blk.GasUsed()))
+	blockMtc.WithLabelValues("MintActions").Set(float64(len(blk.Actions)))
+	return blk, nil
+}
+
+// TipHeight returns tip block's height
+func (cm *forkChain) TipHeight() uint64 {
+	return cm.head.Height()
+}
+
+// TipHash returns tip block's hash
+func (cm *forkChain) TipHash() hash.Hash256 {
+	return cm.tipHash()
+}
+
+// ChainAddress returns chain address on parent chain, the root chain return empty.
+func (cm *forkChain) ChainAddress() string {
+	return cm.bc.ChainAddress()
+}
+
+// StateReader returns the state reader
+func (cm *forkChain) StateReader() protocol.StateReader {
+	return cm.sr
+}
+
+// ValidateBlock validates a new block before adding it to the blockchain
+func (cm *forkChain) ValidateBlock(blk *block.Block) error {
+	err := cm.bc.ValidateBlock(blk)
+	if err != nil {
+		return err
+	}
+	if err = cm.pool.AddBlock(blk); err != nil {
+		blkHash := blk.HashBlock()
+		log.L().Error("failed to add proposal", zap.Error(err), zap.Uint64("height", blk.Height()), log.Hex("hash", blkHash[:]))
+	}
+	return nil
+}
+
+// CommitBlock validates and appends a block to the chain
+func (cm *forkChain) CommitBlock(blk *block.Block) error {
+	if err := cm.bc.CommitBlock(blk); err != nil {
+		return err
+	}
+	if err := cm.bbf.ReceiveBlock(blk); err != nil {
+		blkHash := blk.HashBlock()
+		log.L().Error("failed to receive block", zap.Error(err), zap.Uint64("height", blk.Height()), log.Hex("hash", blkHash[:]))
+	}
+	if err := cm.pool.ReceiveBlock(blk); err != nil {
+		blkHash := blk.HashBlock()
+		log.L().Error("failed to receive block", zap.Error(err), zap.Uint64("height", blk.Height()), log.Hex("hash", blkHash[:]))
+	}
+	return nil
+}
+
+func (cm *chainManager) Start(ctx context.Context) error {
+	head, err := cm.bc.BlockHeaderByHeight(cm.bc.TipHeight())
+	if err != nil {
+		return errors.Wrap(err, "failed to get the head block")
+	}
+	sr, err := cm.srf.StateReaderAt(head)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create state reader at %d, hash %x", head.Height(), head.HashBlock())
+	}
+	cm.forkChain.head = head
+	cm.forkChain.sr = sr
+	cm.pool.Init(cm.forkChain.TipHash())
+	return nil
+}
+
+func (cm *chainManager) Final() ForkChain {
+	return cm.forkChain
+}
+
+// Fork creates a new chain manager with the given hash
+func (cm *chainManager) Fork(hash hash.Hash256) (ForkChain, error) {
```

Comment:
`Fork` is only implemented in `chainManager`, therefore, we may have an instance of `ChainManager` in `forkChain`. then, we don't need to pass in `bc`, `timeFactory`, `bbf`, and `pool` in `forkChain`.